### ISSUE TITLE
[prefer-nullish-coalescing] add option to only flag where behavior would be the same

### ISSUE THREAD

**Issue by felixfbecker** [typescript-eslint/typescript-eslint#1265]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/prefer-nullish-coalescing": "error"
  }
}
```

```TS
function fn(bar?: string) {
	const foo = bar || 'defaultval' // not an error
}

interface SomeObject {
	someProp: boolean
}
function fn2(bar?: SomeObject) {
	const foo = bar || { someProp: true } // error
}
```

**Expected Result**

This should not trigger a rule error, because using `??` would be a different behavior: It would not apply the default value if `bar` is an empty string.

I would expect there to be an option so that it only flags cases where it can confidently say that there is no change in behavior, like when the type is an object. Maybe something like `ignoreTypes: ["number", "string", "boolean"]`.

**Actual Result**

It flags uses of `||` where the behavior is different than `??`, i.e. with strings, booleans and numbers.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.9.0` |
| `@typescript-eslint/parser`        | `2.9.0` |
| `TypeScript`                       | `3.7.2` |
| `ESLint`                           | `6.6.0` |
| `node`                             | `12.11.0` |
| `npm`                              | `6.13.1` |

**Comment by bradzacher** [comment]
I thought about this, but the problem is that this rule would pretty much never fire if it were to be disabled for falsey primitives. Taking `string`, `number`, `boolean`, and `bigint` out of the equation leaves you with just `object` and `symbol`.

From my survey of the usage of nullish coalescing on a large codebase (FB's codebase): it's rare that I see people use chained logical operators for things other than the falsey primitives and arrays. I think the reason for this is that it's rare that you've got an object that is easy to construct with defaults, but it's easy to default a falsey primitive. Also generally nullish objects are handled instead via optional chaining, and people default the nullish output using nullish coalescing (i.e. still operating on falsey primitives).

Could potentially convert the fixer to a suggestion in these cases, so it is not automatically applied.

---

As an aside (this is a personal preferece) - you should endeavour to not rely upon the falsey behaviour of the primitives. Being explicit and showing clear intention for both null and falsey values is much clearer, and leads to a safer codebase.

This is why flow has a compiler error built into strict mode, which prevents you from using the falsey behaviour.

**Comment by felixfbecker** [comment]
> As an aside (this is a personal preferece) - you should endeavour to not rely upon the falsey behaviour of the primitives. Being explicit and showing clear intention for both null and falsey values is much clearer, and leads to a safer codebase.

I agree - the problem is that adding this rule to a codebase, there are countless violations, and changing them is dangerous as you cannot prove that no part relies on this behavior. So I'd rather have at least a subset checked at first and slowly transition to using it for everything.

Also, I think in any case where the input comes from outside your system, e.g. an API request, you usually _do_ want to treat an empty string as if a string was not provided (`undefined`) and _doing so_ leads to a safer codebase. I agree for anything within the system boundaries though.

> Could potentially convert the fixer to a suggestion in these cases, so it is not automatically applied.

I think it should definitely be a suggestion as it changes runtime behavior.
